[Clang][C++26] Implement Pack Indexing (P2662R3).

cor3ntin opened the pull request:
https://isocpp.org/files/papers/P2662R3.pdf

Because there is a slight chance the syntax might change slightly (see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2994r0.html), the feature is not exposed in other language modes.

cor3ntin committed:
[Clang][C++26] Implement Pack Indexing (P2662R3).

https://isocpp.org/files/papers/P2662R3.pdf

Because there is a slight chance the syntax might change slightly
(see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2994r0.html),
the feature is not exposed in other language modes.

erichkeane commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -1061,6 +1070,59 @@ ExprResult Sema::ActOnSizeofParameterPackExpr(Scope *S,
 }
 
+static bool isParameterPack(Expr *PackExpression) {
+  if (auto D = dyn_cast<DeclRefExpr>(PackExpression); D) {
```
```suggestion
  if (auto *D = dyn_cast<DeclRefExpr>(PackExpression); D) {
```

erichkeane reviewed (commented):
A LOT of stuff here... probably going to need a few more cycles on this to completely grok it.  I didn't spend much time on tests, but would love if we could come up with some 'common patterns' (particularly any that others have sent you as a 'look at this cool thing i can do with this!) to put in tests.

cor3ntin commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -4193,6 +4194,13 @@ void CXXNameMangler::mangleType(const PackExpansionType *T) {
   mangleType(T->getPattern());
 }
 
+void CXXNameMangler::mangleType(const PackIndexingType *T) {
```
This is a good question: do we need specific mangling, and if so, how do we process?
I don't think so, but then again I'm always confused that we sometime mangle dependent names

cor3ntin replied:
I did open an issue there https://github.com/itanium-cxx-abi/cxx-abi/issues/175
@zygoloid

cor3ntin replied:
@rnk I was told that you could help with the microsoft abi for this. thanks!

cor3ntin replied:
I did ping Cameron, who will get back to me,
I think we should land this (in 19) even if the abi is not stable - and wait for it to be to set the feature test macro

cor3ntin commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -1061,6 +1070,59 @@ ExprResult Sema::ActOnSizeofParameterPackExpr(Scope *S,
+    Index = Value.getExtValue();
+  }
+
+  if (Index && (!ExpandedExprs.empty() || EmptyPack)) {
```
I think once we have an index, it's always a ConstantExpr so presumably we could extract the value from that.
I suppose keeping the expression is best if people want to do funny things with matchers?

cor3ntin commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -5697,6 +5698,39 @@ QualType ASTContext::getDecltypeType(Expr *e, QualType UnderlyingType) const {
+                                         ArrayRef<QualType> Expansions,
+                                         int Index) const {
+  QualType Canonical;
+  if (FullyExpanded && Index != -1) {
```
given ```cpp

auto f(auto... a) -> decltype(a...[0]);

auto g(auto b...) {
   return f<int>(b...);
}
```

We end up in a situation where we have some expansion for `a...` because explicit parameters (int) are substituted, but we still have a pack. 

We seem to use "Partially substituted" in some other places, would that be more clear?

erichkeane replied:
Yeah, I think 'partially substituted' or 'fully substituted' means what we want here.

cor3ntin replied:
Done!

cor3ntin committed:
Address some comments from Erich

philnik777 commented on clang/lib/Parse/ParseCXXInlineMethods.cpp:
```diff
@@ -978,6 +978,20 @@ bool Parser::ConsumeAndStoreFunctionPrologue(CachedTokens &Toks) {
         break;
       }
+      // Pack indexing
+      if (getLangOpts().CPlusPlus26 && Tok.is(tok::ellipsis) &&
```
FWIW it would be really sweet if we could get this in older language modes to implement `std::tuple`. I don't know whether the current version is ABI compatible, but IMO it would definitely be worth looking into it.

erichkeane replied:
I'd suggested offline that we bring @ldionne in and see if he can make a case for doing so.  However, this IS somewhat of a 'breaking change' for a minor case, so we have to balance that.

ldionne replied:
@philnik777 This would require GCC to also support that as an extension in C++11 (and above). Otherwise we wouldn't be able to use it in our tuple implementation. If we had such an extension, I suspect compilation times for any tuple-heavy code would go down quite a bit. In fact the most interesting part is that we might be able to simplify how we store elements inside the tuple -- right now we have to use some kind of base class trick because that's how we manage to perform indexing afterwards, but that's expensive.

erichkeane replied:
Would a change like that be possible without being an ABI break?  I might bring @AaronBallman into this here discussion and see what we think here.  I see value in enabling it here and hoping GCC does someday, but not sure I'm in favor of it, unless this is something you could use ANYWAY without their support.

Another thing is perhaps we could ask @jicama what GCC intends to do here (or at least for his thoughts/see if we can influence them)?  @jwakely might have something to say as well.

ldionne replied:
I think it might be possible to maintain the same ABI we have right now while changing how it's implemented, yes. It probably requires a bit of thought but it's probably possible. But there's certainly no point in any of this if GCC doesn't add this extension, from my point of view at least.

philnik777 replied:
> @philnik777 This would require GCC to also support that as an extension in C++11 (and above). Otherwise we wouldn't be able to use it in our tuple implementation.

Yeah, obviously. But if one implementation adds it, it becomes a lot more likely that others do it too. I also suspect that libstdc++ may also really like this.



> Would a change like that be possible without being an ABI break?

I don't know, but IMO it would even be worth just adding it in ABIv2 if we can get it as an extension. I expect this to bring a _lot_ of improvement in terms of compile times and debug binary size.

cor3ntin commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -1713,6 +1716,11 @@ class ASTContext : public RefCountedBase<ASTContext> {
+  QualType getPackIndexingType(QualType Pattern, Expr *IndexExpr,
+                               bool FullyExpanded = false,
+                               ArrayRef<QualType> Expansions = {},
+                               int Index = -1) const;
```
I think we really want it there.
We need to be able to get to the selected type to determine the canonical type.

I have compromised by removing it from PackIndexingExpr/PackIndexingType, that way it is not serialized

cor3ntin committed:
Address more of Erich's feedback
 - Add a computeDependence function
 - Remove the Index fron the type / expression:
we do keep it around in some of the functions to avoid haviung to recompute the index in multiple places
in short sucession.

cor3ntin committed:
Allow pack indexing as an extension (wip)

cor3ntin committed:
* Support pack indexing in older language modes
 * improve error recovery for ill-formed pack indexing specifiers

This is important because we need to form a correct type for
an invalid pack indexing that we know is a type pack indexing so
that it is recognized as being part of a declaration.

erichkeane commented on clang/include/clang/AST/ExprCXX.h:
```diff
@@ -4344,6 +4344,108 @@ class SizeOfPackExpr final
+
+
+  std::optional<unsigned> getSelectedIndex() const {
+    if (isInstantiationDependent())
```
Is there good reason this isn't an assert?

cor3ntin replied:
It's called from places that don't have another way to check whether an it would be valid.
Now that we don't store an index, we have to look at the expression to establish what the index is.

erichkeane commented on clang/include/clang/AST/ExprCXX.h:
```diff
@@ -4344,6 +4344,108 @@ class SizeOfPackExpr final
+  }
+
+  ArrayRef<Expr *> getExpressions() const {
+    if (TransformedExpressions == 0)
```
Is this test necessary?  Worst case it just gives an empty collection, right?

erichkeane commented on clang/lib/AST/ExprCXX.cpp:
```diff
@@ -1665,6 +1665,43 @@ NonTypeTemplateParmDecl *SubstNonTypeTemplateParmExpr::getParameter() const {
+                                           std::optional<int64_t> Index,
+                                           ArrayRef<Expr *> SubstitutedExprs) {
+  QualType Type;
+  if (Index && !SubstitutedExprs.empty())
```
We probably want to just do `if (Index)` then `assert(!SubstitutedExprs.empty())` is what I meant here.

cor3ntin replied:
We can get into a situation where the index is no longer dependent, but the pack itself still is. This is why there is this check. the type is dependent until both the index and are expanded

erichkeane commented on clang/lib/AST/Type.cpp:
```diff
@@ -3767,6 +3767,57 @@ void DependentDecltypeType::Profile(llvm::FoldingSetNodeID &ID,
+    return std::nullopt;
+  // Should only be not a constant for error recovery.
+  ConstantExpr *CE = dyn_cast<ConstantExpr>(getIndexExpr());
+  if (!CE)
```
Based on the comment, should this be: `if(getIndexExpr().containsErrors()) return std::nullopt;`, then just a `cast`?

cor3ntin replied:
The expression itself is fine, but it may not be a constant. Do we think we should set the error flag instead?

erichkeane replied:
Why wouldn't the expression be a constant expression here?  The comment claims only when it is an Error Recovery situation (which makes it containsErrors).

cor3ntin replied:
```
int i = 0;
using foo = pack...[i];
```

We need to somehow parse that (if we fail to parse it it's going to at best not be classified as a type),
but i is not a constant expression. So it's error recovery in the general sense, not a RecoveryExpr

erichkeane replied:
Ah, thats what I missed.  I thought you meant that this would be a 'containsError' situation.  I think I'm fine then.

cor3ntin replied:
If you have idea for a better comment...

erichkeane replied:
I don't, but I'll think about it and might comment my next time thru.

cor3ntin committed:
ParseIndexedTypeNamePack -> ParsePackIndexingType

cor3ntin commented on clang/include/clang/Basic/TokenKinds.def:
```diff
@@ -836,6 +836,8 @@ ANNOTATION(primary_expr) // annotation for a primary expression, used when
                          // message send
 ANNOTATION(decltype)     // annotation for a decltype expression,
                          // e.g., "decltype(foo.bar())"
+ANNOTATION(indexed_pack_type) // annotation for an indexed pack of type,
```
I did remain a lot of types to be more consistent, let me know what you think

cor3ntin commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -1884,6 +1884,10 @@ class Parser : public CodeCompletionHandler {
   ExprResult tryParseCXXIdExpression(CXXScopeSpec &SS, bool isAddressOfOperand,
                                      Token &Replacement);
+
+  ExprResult tryParseCXXPackIndexingExpression(ExprResult PackIdExpression);
```
I'm not sure what you mean, could you clarify?

cor3ntin commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -1061,6 +1070,59 @@ ExprResult Sema::ActOnSizeofParameterPackExpr(Scope *S,
+  std::optional<int64_t> Index;
+  if (!IndexExpr->isValueDependent() && !IndexExpr->isTypeDependent()) {
+    llvm::APSInt Value(Context.getIntWidth(Context.getSizeType()));
+    // TODO: do we need a new enumerator instead of CCEK_ArrayBound?
```
It can't be a vla

cor3ntin committed:
More cleanups

cor3ntin commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -1196,6 +1196,95 @@ void Parser::AnnotateExistingDecltypeSpecifier(const DeclSpec &DS,
 }
 
+SourceLocation Parser::ParseIndexedTypeNamePack(DeclSpec &DS) {
+  assert(Tok.isOneOf(tok::annot_indexed_pack_type, tok::identifier) &&
```
When we annotate the token, we go through this function (and once we parse a type, we convert it an annotation), and sometimes we use the type directly (destructor case)

cor3ntin committed:
Add extension warnings tests

cor3ntin committed:
Format

erichkeane reviewed (approved):
No more comments, this looks good to me, and now that we are post release-branch, I think I'm happy to have this merged.

cor3ntin committed:
Rebase

cor3ntin committed:
Fix test portability on windows

cor3ntin committed:
Pack indexing targets Clang 19

cor3ntin commented:
The libcxx test failure are unrelated.

cor3ntin commented:
Fixed in 143b510a8fefb0d203c9ac7e036aa9967dd8a2ef

DamonFool commented:
lldb may also need to be fixed?
```
llvm-project/lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:4074:11: error: enumeration value 'PackIndexing' not handled in switch [-Werror,-Wswitch]
  switch (qual_type->getTypeClass()) {
          ^~~~~~~~~~~~~~~~~~~~~~~~~
llvm-project/lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:4751:11: error: enumeration value 'PackIndexing' not handled in switch [-Werror,-Wswitch]
  switch (qual_type->getTypeClass()) {
          ^~~~~~~~~~~~~~~~~~~~~~~~~
llvm-project/lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:5080:11: error: enumeration value 'PackIndexing' not handled in switch [-Werror,-Wswitch]
  switch (qual_type->getTypeClass()) {
          ^~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
```

cor3ntin commented:
> lldb may also need to be fixed?
> 
> ```
> llvm-project/lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:4074:11: error: enumeration value 'PackIndexing' not handled in switch [-Werror,-Wswitch]
>   switch (qual_type->getTypeClass()) {
>           ^~~~~~~~~~~~~~~~~~~~~~~~~
> llvm-project/lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:4751:11: error: enumeration value 'PackIndexing' not handled in switch [-Werror,-Wswitch]
>   switch (qual_type->getTypeClass()) {
>           ^~~~~~~~~~~~~~~~~~~~~~~~~
> llvm-project/lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:5080:11: error: enumeration value 'PackIndexing' not handled in switch [-Werror,-Wswitch]
>   switch (qual_type->getTypeClass()) {
>           ^~~~~~~~~~~~~~~~~~~~~~~~~
> 3 errors generated.
> ```

Yes. (I missed that, thanks). I'm not familiar with lldb codebase but the fix looks trivial.

cor3ntin commented:
https://github.com/llvm/llvm-project/pull/79695

bevin-hansson reviewed (commented):
We hit some odd behavior in our downstream after this patch.

cor3ntin commented:
@bevin-hansson Thanks for letting me know, I'll look into it
